### PR TITLE
Allow to provide constants to a starlark script

### PR DESCRIPTION
**Pull request by essobedo**
Fix for #8768

## Motivation

We would like to be able to reuse Starlark scripts by providing somehow input parameters

## Modifications:

* Adds a map `constants` to the configuration
* Injects the content of the constants into the `builtins` of the program

### Required for all PRs:

- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

File: plugins/processors/starlark/starlark_test.go [hunk 1]
```diff
@@ -2418,13 +2419,64 @@ def process(metric):
...
+			constants: map[string]interface{}{
+				"max_size":         10,
+				"threshold":        0.75,
+				"default_name":     "Julia",
+				"debug_mode":       true,
+				"supported_values": []interface{}{2, "3"},
+				"supported_entries": map[interface{}]interface{}{
+					2:   "two",
+					"3": "three",
+				},
```
Comment (ssoroka) [inline plugins/processors/starlark/starlark_test.go:2445]: I think we need to test parsing this from TOML to see what the types actually end up being for various map types.

File: plugins/processors/starlark/README.md [hunk 2]
```diff
@@ -182,6 +189,29 @@ def apply(metric):
...
+```toml
+[[processors.starlark]]
+  script = "/usr/local/bin/myscript.star"
+
+  [processors.starlark.constants]
+    somecustomnum = 10
+    somecustomstr = "mycustomfield"
+```
+
+Your script could then use the constants defined in the configuration as next:
```
Comment (Jagularr) [inline plugins/processors/starlark/README.md:207]: grammar nit-pick: replace "the configuration as next" with "the following configuration"

**Review by Jagularr — APPROVED** [review]

File: plugins/processors/starlark/starlark_test.go (hunk 1 above)
Comment (essobedo) [reply plugins/processors/starlark/starlark_test.go:2447]: @ssoroka good catch! Adding a test with config made me realized that the code was buggy.

**Comment by srebhan** [comment]
Hey guys, just a comment/suggestion. I do have some PoC-code for adding templates to starlark. That is you write a template-star-file and fill in the template variables through the config. The template is then processed by a [template engine](github.com/flosch/pongo2) before execution. This way users can exchange templates and adapt to their env without touching the starlark code.

@essobedo Here is some prototype code [code (starlark_templates.go.zip)](https://github.com/influxdata/telegraf/files/5904099/starlark_templates.go.zip) in the hope it's polished to the point of inclusion. It's a standalone test but I'm pretty sure you are able to port it to the plugin.

File: plugins/processors/starlark/field_dict.go [hunk 3]
```diff
@@ -210,17 +211,44 @@ func (i *FieldIterator) Done() {
...
+			sValue, err := asStarlarkValue(iter.Value().Interface())
+			if err != nil {
+				return starlark.None, err
+			}
+			dict.SetKey(sKey, sValue)
+		}
+		return dict, nil
+	case reflect.Float32, reflect.Float64:
+		return starlark.Float(v.Float()), nil
+	case reflect.Int, reflect.Int8, reflect.Int32, reflect.Int64:
```
Comment (ssoroka) [inline plugins/processors/starlark/field_dict.go:244]: missed reflect.Int16

File: plugins/processors/starlark/field_dict.go [hunk 4]
```diff
@@ -210,17 +211,44 @@ func (i *FieldIterator) Done() {
...
+		return starlark.MakeInt64(v.Int()), nil
+	case reflect.Uint, reflect.Uint8, reflect.Uint32, reflect.Uint64:
```
Comment (ssoroka) [inline plugins/processors/starlark/field_dict.go:246]: missed reflect.Uint16

**Review by ssoroka — APPROVED** [review]

**Comment by ssoroka** [comment]
looks good to me.

**Review by reimda — APPROVED** [review]

**Comment by mirkocomparetti-synesis** [comment]
Hello
Is it included in 1.17.3?
That way I can start using it!

Thanks a lot!

**Comment by mirkocomparetti-synesis** [comment]
Thanks a lot. It seems ok in our use case!

**Pull request merged**